adapters/adprime/adprime.go

```diff
@@ -0,0 +1,140 @@
+package adprime
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/mxmCherry/openrtb"
+	"github.com/prebid/prebid-server/adapters"
+	"github.com/prebid/prebid-server/errortypes"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+// AdprimeAdapter struct
+type AdprimeAdapter struct {
+	URI string
+}
+
+// NewAdprimeBidder Initializes the Bidder
+func NewAdprimeBidder(endpoint string) *AdprimeAdapter {
+	return &AdprimeAdapter{
+		URI: endpoint,
+	}
+}
+
+type adprimeParams struct {
+	TagID string `json:"TagID"`
+}
+
+// MakeRequests create bid request for adprime demand
+func (a *AdprimeAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var adprimeExt openrtb_ext.ExtImpAdprime
+	var err error
+
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb.Imp{imp}
+
+		var bidderExt adapters.ExtImpBidder
+		if err = json.Unmarshal(reqCopy.Imp[0].Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if err = json.Unmarshal(bidderExt.Bidder, &adprimeExt); err != nil {
+			errs = append(errs, err)
+			continue
```

Comment:
Given that there's only one parameter in `openrtb_ext.ExtImpAdprime` that we care for (`TagID`), maybe we can optimize a bit using `jsonparser` instead of unmarshalling twice.
```
31   func (a *AdprimeAdapter) MakeRequests(request *openrtb.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
32       var errs []error
33   *--  4 lines: var adprimeExt openrtb_ext.ExtImpAdprime--------------------------------------------------
37
38       reqCopy := *request
39       for _, imp := range request.Imp {
40           reqCopy.Imp = []openrtb.Imp{imp}
41
42 -         var bidderExt adapters.ExtImpBidder
43 -         if err = json.Unmarshal(reqCopy.Imp[0].Ext, &bidderExt); err != nil {
44 -             errs = append(errs, err)
45 -             continue
46 -         }
47 -
48 -         if err = json.Unmarshal(bidderExt.Bidder, &adprimeExt); err != nil {
   +         tagID, err = jsonparser.GetString(reqCopy.Imp[0].Ext, ... , "TagID") // write keys in the ... part
   +         if err != nil {
49               errs = append(errs, err)
50               continue
51           }
52
53 -         reqCopy.Imp[0].TagID = adprimeExt.TagID
   +         reqCopy.Imp[0].TagID = tagID
54
55   *--  5 lines: adapterReq, errors := a.makeRequest(&reqCopy)---------------------------------------------
60       }
61       return adapterRequests, errs
62   }
adapters/adprime/adprime.go
```

Reply:
This one done